packages/component-base/src/virtualizer-iron-list-adapter.js

```diff
@@ -297,16 +312,28 @@ export class IronListAdapter {
 
   _scrollHandler() {
     this._adjustVirtualIndexOffset(this._scrollTop - (this.__previousScrollTop || 0));
+    const delta = this.scrollTarget.scrollTop - this._scrollPosition;
 
     super._scrollHandler();
 
     if (this._physicalCount !== 0) {
-      // After running super._scrollHandler, fix _virtualStart to workaround an iron-list issue.
-      // See https://github.com/vaadin/web-components/issues/1691
-      const reusables = this._getReusables(true);
-      this._physicalTop = reusables.physicalTop;
-      this._virtualStart += reusables.indexes.length;
-      this._physicalStart += reusables.indexes.length;
+      const isScrollingDown = delta >= 0;
+      const reusables = this._getReusables(!isScrollingDown);
+
+      if (reusables.indexes.length) {
+        // After running super._scrollHandler, fix internal properties to workaround an iron-list issue.
+        // See https://github.com/vaadin/web-components/issues/1691
+        this._physicalTop = reusables.physicalTop;
+
+        if (isScrollingDown) {
+          this._virtualStart -= reusables.indexes.length;
+          this._physicalStart -= reusables.indexes.length;
+        } else {
+          this._virtualStart += reusables.indexes.length;
+          this._physicalStart += reusables.indexes.length;
+        }
+        this._resizeHandler();
+      }
```

Comment:
Not directly related to this change: a forward-port of https://github.com/vaadin/vaadin-grid/pull/2217 to fix some tests that started failing